Do not update etc/hosts file for every container

aboch opened the pull request:
For non default network:
- Only containers with a user chosen name will be published in the etc/hosts file of the other containers.
  This behavior is achieved connecting auto-generated name containers to a network via anonymous endpoints. Anonymous endpoints are not published in the network. User named containers are instead connected to the network via regular endpoints, which get published across the network.
- This change does not affect linked containers. Publishing of linked containers will work as today.

Also block linking to containers which are not connected to the default network

HuKeping commented on integration-cli/docker_cli_network_unix_test.go:
```diff
@@ -384,3 +385,59 @@ func (s *DockerNetworkSuite) TestDockerNetworkDriverOptions(c *check.C) {
+	// verify first container etc/hosts file has not changed
+	hosts1post, err := readContainerFileWithExec(cid1, hostsFile)
+	c.Assert(err, checker.IsNil)
+	if !bytes.Equal(hosts1, hosts1post) {
```
It would be better if using the `c.Assert`

aboch replied:
Thanks @HuKeping.
I will make the suggested changes here and below.

HuKeping commented on integration-cli/docker_cli_network_unix_test.go:
```diff
@@ -384,3 +385,59 @@ func (s *DockerNetworkSuite) TestDockerNetworkDriverOptions(c *check.C) {
+	// verify etc/hosts file for first two containers contains the named container entry
+	hosts1post, err = readContainerFileWithExec(cid1, hostsFile)
+	c.Assert(err, checker.IsNil)
+	if !strings.Contains(string(hosts1post), cName) {
```
something like `c.Assert(hosts1post, checker.Contains, cName, check.Commentf(...))`

HuKeping commented on integration-cli/docker_cli_network_unix_test.go:
```diff
@@ -384,3 +385,59 @@ func (s *DockerNetworkSuite) TestDockerNetworkDriverOptions(c *check.C) {
+
+	hosts2post, err := readContainerFileWithExec(cid2, hostsFile)
+	c.Assert(err, checker.IsNil)
+	if !strings.Contains(string(hosts2post), cName) {
```
same here

HuKeping commented on integration-cli/docker_cli_network_unix_test.go:
```diff
@@ -384,3 +385,59 @@ func (s *DockerNetworkSuite) TestDockerNetworkDriverOptions(c *check.C) {
+	dockerCmd(c, "stop", cid3)
+	hosts1post, err = readContainerFileWithExec(cid1, hostsFile)
+	c.Assert(err, checker.IsNil)
+	if !bytes.Equal(hosts1, hosts1post) {
```
same here and below

HuKeping commented on integration-cli/docker_cli_network_unix_test.go:
```diff
@@ -384,3 +385,59 @@ func (s *DockerNetworkSuite) TestDockerNetworkDriverOptions(c *check.C) {
+
+	hosts2post, err = readContainerFileWithExec(cid2, hostsFile)
+	c.Assert(err, checker.IsNil)
+	if !bytes.Equal(hosts2, hosts2post) {
```
:smile_cat:

mavenugo commented:
LGTM. 
this also overrides https://github.com/docker/docker/pull/17229 (due to latest libnetwork vendor-in).

cpuguy83 commented:
Is this to reduce the amount of writes happening?
Seems a bit hackish to add in... I would prefer removing the functionality all-together rather than disabling only sometimes.

If the goal is to limit the number of writes, there could be a psuedo-TTL used for updating hosts files... e.g. collect changes for X seconds and then write.
The hosts file would be X seconds behind (assuming anything happened during that time) but generally accepted with DNS anyway.

cpuguy83 commented:
I think this will be confusing.

mavenugo commented:
@cpuguy83 this is absolutely not to do with the reducing the amount of writes. This is purely to make sure we dont write the auto generated names to /etc/hosts, because it is NOT expected by the user and is not requested by the user. Any container launched with `--name` has a meaning for discovering by name & it is not appropriate to do so for auto-generated name, where we cannot guarantee uniqueness in a multi-host environment. This was tracked under https://github.com/docker/libnetwork/issues/567 and was originally planned to as part of the UX refactoring https://github.com/docker/docker/pull/16645#issue-108793685 but was pushed out due to the back-end changes.

tiborvass commented:
@aboch need a rebase

aboch committed:
Do not update etc/hosts for every container

- Only user named containers will be published into
  other containers' etc/hosts file.
- Also block linking to containers which are not
  connected to the default network

Signed-off-by: Alessandro Boch <aboch@docker.com>

calavera commented:
Why don't we do the opposite this PR actually does? AKA: if there is a name provided by the user, either by the `--name` flag or the field in the api, we add the container, otherwise we do nothing.

calavera commented:
That way we don't need to add that new on-shot field to the container.

calavera commented:
oh, that's what this does, what if we delay the container name assignation after the network is configured? that way we still don't need that new field.

cpuguy83 commented:
@calavera daemon restart

calavera commented:
@cpuguy83 good point, why do we regenerate the networking configuration after every daemon restart? should it be stateful like the rest of the container configuration?

aboch commented:
@calavera 

Yes the drawbacks is that we need a new field to know if this was an explicitly named container or not at network conencion time. Also for daemon restart as @cpuguy83 pointed out.

Going back to the flag name, I am not happy myself with `GeneratedName`. Do you think something like `Anonymous` or `Named` or `ExplicitlyNamed` makes more sense as container common property ?

tiborvass commented:
@aboch I would vote for `ExplicitlyNamed`. @cpuguy83 @calavera wdyt?

calavera commented:
I feel like this should be into the network settings in the container. The problem is not about the name being generated or not, the problem is about the container added to the hosts or not.

aboch commented:
@calavera 

I need to check if that would work. If it does I am fine with it.

But, before discarding the new flag in the container config, do you guys think it may become handy in future to have that container specific information, ExplicitlyNamed, also based on past community requests (which I am not aware of) ?

cpuguy83 commented:
@calavera agreed, and since discovery is not really solved yet, it's difficult to just provide a flag to disable too.

calavera commented:
no, knowing if the name was generated by us or provided by the user is irrelevant, from my point of view.

cpuguy83 commented:
@calavera also agreed

aboch commented:
Thanks @calavera @cpuguy83 @tiborvass 

I moved the property to `network.Settings` as you suggested.
Changed the property name as it is now network specific.

calavera commented:
LGTM

tiborvass commented:
@aboch can you please make sure `--icc=false` does not update at all `/etc/hosts` for `docker0` ?

cpuguy83 commented:
Except for links

tiborvass commented:
Dropping this for 1.9. Feel free to discuss further the implications of making a distinction between user-picked names (via `--name` or API equivalent) and autogenerated names for service discovery.

tiborvass commented:
@aboch the `--icc=false` fix would be welcome though.

calavera commented:
It needs a rebase :recycle:

tiborvass commented:
@calavera @aboch no need for rebase, this is in design discussion.

tiborvass commented:
Okay after rethinking this, we need this for 1.9 because `docker network` is a new feature in 1.9 and if we don't get this in, then we can't change our minds later without breaking people.

icecrime commented:
Ok that only explicitly named containers get added to `/etc/hosts`. LGTM

thaJeztah commented on daemon/network/settings.go:
```diff
@@ -43,4 +43,5 @@ type Settings struct {
 	SandboxKey             string
 	SecondaryIPAddresses   []Address
 	SecondaryIPv6Addresses []Address
+	AnonymousEndpoint      bool
```
Just a nit while reading, and don't know if we actually have a convention for this, but I like names such as `IsXXX / `HasXXX` for boolean values (in this case e.g. `IsAnonymous`)

tiborvass replied:
agreed

aboch replied:
Makes sense. Will make the change.

icecrime commented on daemon/daemon.go:
```diff
@@ -479,6 +479,7 @@ func (daemon *Daemon) newContainer(name string, config *runconfig.Config, imgID
 	var (
 		id  string
 		err error
+		ane = name == ""
```
Not ideal for readability.

aboch replied:
Did not want to reset indent of previous vars. Being a function's internal var thought it was ok also because it is getting assigned to `IsAnonymousEndpoint`, self explicative.

If you think it is still better to change it, any suggestion ? I can only think of `isAnonymous` or `nameNotSpecified`

thaJeztah replied:
isAnon(ymous) sgtm, but agreed that the `IsAnonymousEndpoint` already helps

aboch replied:
At the function level, we are really checking if a name was not passed for the container. Container will never be anonymous. I should then probably choose `nameNotSpecified` or `noExplicitName`, `nonexplicitlyNamed` in this context. I do not like they are so long...

thaJeztah commented on integration-cli/docker_cli_network_unix_test.go:
```diff
@@ -485,3 +485,58 @@ func (s *DockerDaemonSuite) TestDockerNetworkNoDiscoveryDefaultBridgeNetwork(c *
+	dockerCmd(c, "stop", cid3)
+	hosts1post, err = readContainerFileWithExec(cid1, hostsFile)
+	c.Assert(err, checker.IsNil)
+	c.Assert(string(hosts1), checker.Equals, string(hosts1post),
```
can we guarantee the file is exactly the same as the original (including the order of entries, and whitespace)? Or should we check the entries are "no longer in there"?

thaJeztah replied:
(I don't know how the code works that restores/saves the file, so I may just see problems that are no problems :smile:)

aboch replied:
@thaJeztah The checks actually checks if the `post` file is exactly equal to the original one, byte by byte.

aboch replied:
@thaJeztah Reading again your comment I think you were rather asking whether it is legitimate checking the equality of the pre and post operation files via a raw byte by byte equality check, whether it is too restrictive.

Answer is yes, based on the way libnetwork constructs these file, the raw check is fine.

thaJeztah replied:
Yes, that's what I meant (sorry if I wasn't clear there); we've had some flaky tests, so trying to prevent that from happening.

Thanks for confirming :+1:

tiborvass commented:
For the record, CI UI is stuck, but CI itself is green:
https://jenkins.dockerproject.org/job/Docker-PRs-experimental/9801/console
https://jenkins.dockerproject.org/job/Docker-PRs/18566/console
https://jenkins.dockerproject.org/job/Docker-PRs-userns/946/console
https://jenkins.dockerproject.org/job/Windows-PRs/15797/console

@aboch I tested this PR, we shouldn't allow linking across networks for now. `--link` should be only for the default `bridge0` network. Note that `docker run --net bridge2 --link somecontainer:foo busybox` works, but `docker run --link container_on_bridge2:foo busybox` does not.

aboch commented:
Thanks @tiborvass, I am working at a fix

aboch commented:
@tiborvass 
Pushed the new changes and test. PTAL

LK4D4 commented on daemon/daemon.go:
```diff
@@ -477,8 +477,9 @@ func (daemon *Daemon) getEntrypointAndArgs(configEntrypoint *stringutils.StrSlic
-		err error
+		id             string
+		err            error
+		noExplicitName = name == ""
```
Just put it where it needed, there is no point to create var so early.

aboch replied:
I cannot do that.
`name` gets overwritten at line 483

LK4D4 replied:
ok, got it

LK4D4 commented on daemon/daemon.go:
```diff
@@ -495,7 +496,7 @@ func (daemon *Daemon) newContainer(name string, config *runconfig.Config, imgID
 	base.hostConfig = &runconfig.HostConfig{}
 	base.ImageID = imgID
-	base.NetworkSettings = &network.Settings{}
+	base.NetworkSettings = &network.Settings{IsAnonymousEndpoint: noExplicitName}
```
```
base.NetworkSettings = &network.Settings{IsAnonymousEndpoint: name == ""}
```

aboch replied:
Please see comment above

dave-tucker commented:
hey guys. Shouldn't `--hostname` be written in to the container instead of `--name`.
I thought `--name` was a heplful replacement for the container ID
`--hostname` actually indicate what I want my container to be called from a networking standpoint

this is important.
- you can't have containers with the same name
- you CAN have two containers with the same hostname, it's up to linux to decide which one you hit when you resolv it's name

The latter is how compose scale works

tiborvass commented:
LGTM

tiborvass commented:
@dave-tucker we can discuss `--hostname` for 1.10.

dave-tucker commented:
@tiborvass thanks tibor. i'll open an issue so we can discuss for 1.10
